File: flink-demo/flinkJavaCode/src/main/java/com/yw/flink/example/javacases/case04_sink/Case02_JdbcSinkExactlyOnce.java
//package com.yw.flink.example.javacases.case04_sink;
//
//import com.yw.flink.example.StationLog;
//import com.mysql.jdbc.jdbc2.optional.MysqlXADataSource;
//import org.apache.flink.api.common.functions.MapFunction;
//import org.apache.flink.connector.jdbc.JdbcExactlyOnceOptions;
//import org.apache.flink.connector.jdbc.JdbcExecutionOptions;
//import org.apache.flink.connector.jdbc.JdbcSink;
//import org.apache.flink.connector.jdbc.JdbcStatementBuilder;
//import org.apache.flink.streaming.api.datastream.DataStreamSource;
//import org.apache.flink.streaming.api.datastream.SingleOutputStreamOperator;
//import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
//import org.apache.flink.streaming.api.functions.sink.SinkFunction;
//import org.apache.flink.util.function.SerializableSupplier;
//
//import javax.sql.XADataSource;
//import java.sql.PreparedStatement;
//import java.sql.SQLException;
//
///**
// * Flink 写出JDBC exactly-once 实现
// */
//public class Case02_JdbcSinkExactlyOnce {
//    public static void main(String[] args) throws Exception {
//        StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
//        //设置checkpoint
//        env.enableCheckpointing(1000);
//
//        //socket :005,188,187,busy,5000,50
//        DataStreamSource<String> ds = env.socketTextStream("node5", 9999);
//        //对ds进行转换
//        SingleOutputStreamOperator<StationLog> ds2 = ds.map(new MapFunction<String, StationLog>() {
//            @Override
//            public StationLog map(String line) throws Exception {
//                String[] split = line.split(",");
//                String sid = split[0];
//                String callOut = split[1];
//                String callIn = split[2];
//                String callType = split[3];
//                Long callTime = Long.valueOf(split[4]);
//                Long duration = Long.valueOf(split[5]);
//                return new StationLog(sid, callOut, callIn, callType, callTime, duration);
//            }
//        });
//
//        //准备JDBC Sink 对象
//        /**
//         *
//         * CREATE TABLE `station_log` (
//         *   `sid` varchar(255) DEFAULT NULL,
//         *   `call_out` varchar(255) DEFAULT NULL,
//         *   `call_in` varchar(255) DEFAULT NULL,
//         *   `call_type` varchar(255) DEFAULT NULL,
//         *   `call_time` bigint(20) DEFAULT NULL,
//         *   `duration` bigint(20) DEFAULT NULL
//         * ) ;
//         *
//         * JdbcSink.sink(
//         * sqlDmlStatement,      // 必须指定，SQL语句
//         * jdbcStatementBuilder, // 必须指定，给SQL语句设置参数
//         * jdbcExecutionOptions, // 可选，指定写出参数，如：提交周期、提交批次大小、重试时间，建议指定。
//         * jdbcConnectionOptions // 必须指定，数据库连接参数
//         * );
//         */
//        SinkFunction<StationLog> jdbcSink = JdbcSink.exactlyOnceSink(
//                "insert into station_log (sid,call_out,call_in,call_type,call_time,duration) values (?,?,?,?,?,?)",
//                new JdbcStatementBuilder<StationLog>() {
//                    @Override
//                    public void accept(PreparedStatement pst, StationLog stationLog) throws SQLException {
//                        pst.setString(1, stationLog.sid);
//                        pst.setString(2, stationLog.callOut);
//                        pst.setString(3, stationLog.callIn);
//                        pst.setString(4, stationLog.callType);
//                        pst.setLong(5, stationLog.callTime);
//                        pst.setLong(6, stationLog.duration);
//                    }
//                },
//                JdbcExecutionOptions.builder()
//                        //向数据库提交批次的间隔，默认0
//                        .withBatchIntervalMs(1000)
//                        //数据库提交的批次大小，默认500
//                        .withBatchSize(1000)
//                        //连接重试次数，默认3
//                        .withMaxRetries(0)
//                        .build(),
//                JdbcExactlyOnceOptions.builder()
//                        //允许连接使用事务
//                        .withTransactionPerConnection(true)
//                        .build(),
//                new SerializableSupplier<XADataSource>() {
//                    @Override
//                    public XADataSource get() {
//                        MysqlXADataSource xaDataSource = new MysqlXADataSource();
//                        xaDataSource.setURL("jdbc:mysql://node2:3306/mydb?useSSL=false");
//                        xaDataSource.setUser("root");
//                        xaDataSource.setPassword("123456");
//                        return xaDataSource;
//                    }
//                });
//
//        //数据写出到数据库
//        ds2.addSink(jdbcSink);
//        env.execute();
//    }
//}

File: flink-demo/flinkScalaCode/src/main/scala/com/yw/flink/example/scalacases/case04_sink/Case02_JdbcSinkExactlyOnce.scala
package com.yw.flink.example.scalacases.case04_sink

import com.mysql.jdbc.jdbc2.optional.MysqlXADataSource
import com.yw.flink.example.StationLog
import org.apache.flink.connector.jdbc.{JdbcExactlyOnceOptions, JdbcExecutionOptions, JdbcSink, JdbcStatementBuilder}
import org.apache.flink.streaming.api.functions.sink.SinkFunction
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.util.function.SerializableSupplier

import java.sql.PreparedStatement
import javax.sql.XADataSource

/**
  * Flink JDBC Exactly-once Sink
  */
object Case02_JdbcSinkExactlyOnce {
  def main(args: Array[String]): Unit = {
    val env: StreamExecutionEnvironment = StreamExecutionEnvironment.getExecutionEnvironment
    env.enableCheckpointing(1000)
    import org.apache.flink.streaming.api.scala._
    //003,186,188,busy,3000,30
    val ds1: DataStream[String] = env.socketTextStream("node5", 9999)
    val ds2: DataStream[StationLog] = ds1.map(one => {
      val split: Array[String] = one.split(",")
      val sid: String = split(0)
      val callOut: String = split(1)
      val callIn: String = split(2)
      val callType: String = split(3)
      val callTime: Long = split(4).toLong
      val duration: Long = split(5).toLong
      StationLog(sid, callOut, callIn, callType, callTime, duration)
    })

    //准备JDBC 写出的对象
    /**
      * JdbcSink.sink(
      * sqlDmlStatement,      // 必须指定，SQL语句
      * jdbcStatementBuilder, // 必须指定，给SQL语句设置参数
      * jdbcExecutionOptions, // 可选，指定写出参数，如：提交周期、提交批次大小、重试时间，建议指定。
      * jdbcConnectionOptions // 必须指定，数据库连接参数
      * );
      */

    val jdbcSink: SinkFunction[StationLog] = JdbcSink.exactlyOnceSink[StationLog](
      "insert into station_log values (?,?,?,?,?,?)",
      new JdbcStatementBuilder[StationLog] {
        override def accept(pst: PreparedStatement, station: StationLog): Unit = {
          pst.setString(1, station.sid)
          pst.setString(2, station.callOut)
          pst.setString(3, station.callIn)
          pst.setString(4, station.callType)
          pst.setLong(5, station.callTime)
          pst.setLong(6, station.duration)
        }
      },
      JdbcExecutionOptions.builder()
        .withBatchSize(500)
        .withBatchIntervalMs(200)
        .withMaxRetries(0).build(),
      JdbcExactlyOnceOptions.builder().withTransactionPerConnection(true).build(),

      new SerializableSupplier[XADataSource]() {
        override def get: XADataSource = {
          val xaDataSource: MysqlXADataSource = new MysqlXADataSource
          xaDataSource.setURL("jdbc:mysql://node2:3306/mydb?useSSL=false")
          xaDataSource.setUser("root")
          xaDataSource.setPassword("123456")
          return xaDataSource
        }
      }
    )

    //将数据写出到JDBC中
    ds2.addSink(jdbcSink)
    env.execute()
  }
}
